code/datums/explosion.dm

```diff
@@ -177,24 +237,68 @@
 		UnexplodeT.explosion_level = 0
 	exploded_this_tick.Cut()
 
-	var/took = (world.timeofday-start)/10
-	//You need to press the DebugGame verb to see these now....they were getting annoying and we've collected a fair bit of data. Just -test- changes  to explosion code using this please so we can compare
+	var/took = (REALTIMEOFDAY - started_at) / 10
+
+	//You need to press the DebugGame verb to see these now....they were getting annoying and we've collected a fair bit of data. Just -test- changes to explosion code using this please so we can compare
 	if(GLOB.Debug2)
 		log_world("## DEBUG: Explosion([x0],[y0],[z0])(d[devastation_range],h[heavy_impact_range],l[light_impact_range]): Took [took] seconds.")
 
-	//Machines which report explosions.
-	for(var/array in GLOB.doppler_arrays)
-		var/obj/machinery/doppler_array/A = array
-		A.sense_explosion(epicenter,devastation_range,heavy_impact_range,light_impact_range,took,orig_dev_range,orig_heavy_range,orig_light_range)
+	if(!stopped)	//if we aren't in a hurry
+		//Machines which report explosions.
+		for(var/array in GLOB.doppler_arrays)
+			var/obj/machinery/doppler_array/A = array
+			A.sense_explosion(epicenter, devastation_range, heavy_impact_range, light_impact_range, took,orig_dev_range, orig_heavy_range, orig_light_range)
+
+	++stopped
+	qdel(src)
+
+#undef EX_PREPROCESS_EXIT_CHECK
+#undef EX_PREPROCESS_CHECK_TICK
 
-	return 1
+//asyncly populate the affected_turfs list
+/datum/explosion/proc/GatherSpiralTurfs(range, turf/epicenter)
+	set waitfor = FALSE
+	. = list()
+	spiral_range_turfs(range, epicenter, outlist = ., tick_checked = TRUE)
+	++stopped
 
+/datum/explosion/proc/CaculateExplosionBlock(list/affected_turfs)
+	set waitfor = FALSE
 
+	. = list()
+	var/processed = 0
+	while(!stopped && running)
+		var/I 
+		for(I in (processed + 1) to affected_turfs.len) // we cache the explosion block rating of every turf in the explosion area
+			var/turf/T = affected_turfs[I]
+			var/current_exp_block = T.density ? T.explosion_block : 0
 
-/proc/secondaryexplosion(turf/epicenter, range)
-	for(var/turf/tile in spiral_range_turfs(range, epicenter))
-		tile.ex_act(2)
+			for(var/obj/machinery/door/D in T)
+				if(D.density)
+					current_exp_block += D.explosion_block
 
+			for(var/obj/structure/window/W in T)
```

Comment:
@Iamgoofball typeless loops only apply when the list is full of only one type you dip